app/src/main/java/com/chouten/app/domain/use_case/module_use_cases/AddModuleUseCase.kt

```diff
@@ -0,0 +1,370 @@
+                        semverCompar("0.0.0", it.second.version)
+                    } catch (e: IllegalArgumentException) {
+                        // The existing module has an invalid version
+                        // We can make a note of it and let the module be installed
+                        log("Module ${it.second.name} (${it.second.id}) has an invalid version (${it.second.version})")
+                        1
+                    }
+                }
+
+                when (ret) {
+                    // Module being installed is newer than the existing module
+                    1 -> {
+                        // Delete the old module
+                        if (DocumentFile.fromSingleUri(mContext, it.first)?.delete() == false) {
+                            log("Could not delete module ${it.second.name} (${it.second.id})")
+                        }
+                        log("Updated module ${module.name} (${module.id})")
+                    }
+
+                    // Module being installed is the same version as the existing module
+                    0 -> {
+                        safeException(
+                            IllegalArgumentException("Module ${module.name} (${module.id}) already exists"),
+                            newModuleUri
+                        )
+                    }
+
+                    // Module being installed is older than the existing module
+                    else -> {
+                        safeException(
+                            IllegalArgumentException("Module ${module.name} (${module.id}) is older than the existing module"),
+                            newModuleUri
+                        )
+                    }
+                }
+            }
+        }
+
+        // Rename the module folder to the module name
+
+        /**
+         * A DocumentFile of newModuleUri
+         */
+        val moduleFolder = DocumentFile.fromTreeUri(
+            mContext, newModuleUri.toString().removeSuffix("/children").toUri()
+        ) ?: safeException(IOException("Could not get module folder"), newModuleUri)
+
+        // Not being able to rename the module folder is not a critical error
+        // So we will just log it and continue
+        try {
+            if (!moduleFolder.renameTo(module.name)) {
+                log("Could not rename module folder $newModuleUri to ${module.name}")
+            } else log("Successfully added module ${module.name} (${module.id})")
+        } catch (e: Exception) {
+            log("Could not rename module folder $newModuleUri to ${module.name}")
+        }
+    }
+
+    /**
+     * Compare two semver versions
+     * @param v1 The first version
+     * @param v2 The second version
+     * @return 1 if the first version is newer, -1 if the second version is newer, 0 if the versions are the same
+     * @throws IllegalArgumentException if the versions are invalid
+     */
+    private fun semverCompar(v1: String, v2: String): Int {
```

Comment:
```suggestion
     * Compare two semVer versions
     * @param v1 The first version
     * @param v2 The second version
     * @return 1 if the first version is newer, -1 if the second version is newer, 0 if the versions are the same
     * @throws IllegalArgumentException if the versions are invalid
     */
    private fun semVerCompar(v1: String, v2: String): Int {
```
_not sure on this one, up to you_